File: app/src/lib/game/computer/modules/whot-intelligence.ts
import { SUIT_VALUES } from '../../types.js';
import type { Candidate, ModuleContext, ScoringModule } from '../types.js';

const MAX_PER_SHAPE = SUIT_VALUES.length; // 12

// When playing a Whot, call the shape that opponents are least likely to hold.
// Fewer remaining in a shape → it's been mostly played → opponents probably don't have it.
export const whotIntelligence: ScoringModule = (
	candidate: Candidate,
	ctx: ModuleContext
): number => {
	if (candidate.kind !== 'play-whot') return 0;

	const remaining = ctx.shapeRemaining[candidate.calledShape];
	// Scale 0–8: calling a depleted shape is strongly preferred
	return ((MAX_PER_SHAPE - remaining) / MAX_PER_SHAPE) * 8;
};

File: server/src/game/ai/modules/whot_intelligence.rs
use crate::game::{
    ai::types::{Candidate, ModuleContext},
    engine::CARDS_PER_SHAPE,
};

/// When playing a Whot (or declaring an opening Whot's shape), call the shape
/// opponents are least likely to hold.
pub fn whot_intelligence(candidate: &Candidate, ctx: &ModuleContext<'_>) -> f64 {
    let called_shape = match candidate {
        Candidate::PlayWhot { called_shape } => called_shape,
        Candidate::CallShape { shape } => shape,
        _ => return 0.0,
    };

    // For an actual Whot play, don't bias toward it when a suit play exists — let
    // suit cards win. An opening CallShape has no alternative, so always score it.
    if matches!(candidate, Candidate::PlayWhot { .. })
        && ctx
            .candidates
            .iter()
            .any(|c| matches!(c, Candidate::PlaySuit { .. }))
    {
        return 0.0;
    }

    let remaining = *ctx.shape_remaining.get(called_shape).unwrap_or(&0) as f64;
    let scale = 1.0 / (ctx.state.seats.len() as f64 - 1.0).max(1.0);
    ((CARDS_PER_SHAPE as f64 - remaining) / CARDS_PER_SHAPE as f64) * 15.0 * scale
}
